hack/lib/util/trap.sh

```diff
@@ -0,0 +1,113 @@
+#!/bin/bash
+#
+# This library defines the trap handlers for the ERR and EXIT signals. Any new handler for these signals
+# must be added to these handlers and activated by the environment variable mechanism that the rest use.
+# These functions ensure that no handler can ever alter the exit code that was emitted by a command
+# in a test script.
+
+# This script assumes ${OS_ROOT} is set
+source "${OS_ROOT}/hack/lib/util/misc.sh"
+source "${OS_ROOT}/hack/lib/log/stacktrace.sh"
+
+# os::util::trap::init initializes the privileged handlers for the ERR and EXIT signals if they haven't
+# been registered already
+#
+# Globals:
+#  None
+# Arguments:
+#  None
+# Returns:
+#  None
+function os::util::trap::init() {
+    os::util::trap::init_err
+    os::util::trap::init_exit
+}
+
+# os::util::trap::init_err initializes the privileged handler for the ERR signal if it hasn't
+# been registered already
+#
+# Globals:
+#  None
+# Arguments:
+#  None
+# Returns:
+#  None
+function os::util::trap::init_err() {
+    if ! trap -p ERR | grep -q 'os::util::trap::err_handler'; then
+        trap 'os::util::trap::err_handler;' ERR
+    fi
+}
+
+# os::util::trap::init_exit initializes the privileged handler for the EXIT signal if it hasn't
+# been registered already
+#
+# Globals:
+#  None
+# Arguments:
+#  None
+# Returns:
+#  None
+function os::util::trap::init_exit() {
+    if ! trap -p EXIT | grep -q 'os::util::trap::exit_handler'; then
+        trap 'os::util::trap::exit_handler;' EXIT
+    fi
+}
+
+# os::util::trap::err_handler is the handler for the ERR signal.
+# 
+# Globals:
+#  - OS_USE_STACKTRACE
+#  - OS_TRAP_DEBUG
+# Arguments:
+#  None
+# Returns:
+#  - returns original return code, allows privileged handler to exit if necessary
+function os::util::trap::err_handler() {
+    local -r return_code=$?
+    local -r last_command="${BASH_COMMAND}"
+
+    if set +o | grep -q '\-o errexit'; then
```

Comment:
I'd use `grep -q -e '-foo'` instead of `grep -q '\-foo'`, but I guess it's a stylistic preference.